src/utils/csvParser/parser.ts

```diff
@@ -0,0 +1,611 @@
+    this.metadataTrimSet = metadataTrimSet ?? ''
+    this.resetMode = resetMode || 'none'
+    this.timeFunc = timeFunc ?? (() => new Date())
+
+    if (!this.headerRowCount && !this.columnNames.length) {
+      throw new Error(
+        '`headerRowCount` cannot be 0 if `columnNames` is not specified'
+      )
+    }
+
+    if (this.delimiter.length > 1) {
+      throw new Error(
+        `delimiter must be a single character, got: ${this.delimiter}`
+      )
+    }
+
+    // we might consider removing this configuration option and forcing people to use # as comment
+    // let's see how it goes in the wild
+    if (this.comment.length > 1) {
+      throw new Error(
+        `comment must be a single character, got: ${this.comment}`
+      )
+    }
+
+    if (
+      this.columnNames.length &&
+      this.columnTypes.length &&
+      this.columnNames.length !== this.columnTypes.length
+    ) {
+      throw new Error("columnNames field count doesn't match with columnTypes")
+    }
+
+    if (!['none', 'always'].includes(this.resetMode)) {
+      throw new Error(
+        `expected "none" or "always" but got unknown reset mode ${this.resetMode}`
+      )
+    }
+  }
+}
+
+export class CSVParser {
+  private timeFunc: TimeFunc = () => new Date()
+  private gotColumnNames = false
+  private gotInitialColumnNames = false
+  private remainingSkipRows = 0
+  private remainingHeaderRows = 0
+  private remainingMetadataRows = 0
+  public metadataTags: Record<string, string> = {}
+  public metadataSeparatorList: string[] = []
+  config: Config
+
+  // InitFromConfig
+  constructor(config?: Partial<ParserConfig>) {
+    this.config = new Config(
+      config?.columnNames,
+      config?.columnTypes,
+      config?.comment,
+      config?.defaultTags,
+      config?.delimiter,
+      config?.headerRowCount,
+      config?.measurementColumn,
+      config?.metricName,
+      config?.skipColumns,
+      config?.skipRows,
+      config?.tagColumns,
+      config?.timestampColumn,
+      config?.timestampFormat,
+      config?.timezone,
+      config?.trimSpace,
+      config?.skipValues,
+      config?.skipErrors,
+      config?.metadataRows,
+      config?.metadataSeparators,
+      config?.metadataTrimSet,
+      config?.resetMode,
+      config?.timeFunc
+    )
+
+    this.gotInitialColumnNames = !!this.config.columnNames.length
+
+    this.initializeMetadataSeparator()
+    this.reset()
+  }
+
+  async parse(file: LocalFile) {
+    // Reset the parser according to the specified mode
+    if (this.config.resetMode === 'always') {
+      this.reset()
+    }
+
+    return this.parseCSV(file)
+  }
+
+  async parseLine(line: LocalFile) {
+    if (!line) {
+      if (this.remainingSkipRows > 0) {
+        this.remainingSkipRows--
+        throw EOFError
+      }
+      if (this.remainingMetadataRows > 0) {
+        this.remainingMetadataRows--
+        throw EOFError
+      }
+    }
+
+    const metrics = await this.parseCSV(line)
```

Comment:
can/should this be something like

```ts
const metrics: Metric[] = await ...
```
?

Reply:
we can. But because `this.parseCSV()` is returning metrics with explicit `Metric` type, typescript is able to make that inference. But i don't see the harm in adding it here.